File: 프로그래머스/lv3/43162. 네트워크/네트워크.kt
class Solution {
    fun solution(n: Int, computers: Array<IntArray>): Int {
        val graph = Array(n) { IntArray(n) }
        val visit = BooleanArray(n)

        var i = 0
        computers.forEach {
            graph[i] = it
            i++
        }

        var answer = 0

        for (i in 0 until n) {
            if (!visit[i]) {
                dfs(i, visit, graph)
                answer++
            }
        }
        return answer
    }
}

fun dfs(start: Int, visit: BooleanArray, graph: Array<IntArray>) {
    visit[start] = true
    for (i in graph[start].indices) {
        if (i == start || graph[start][i] != 1) continue
        if (visit[i]) continue
        dfs(i, visit, graph)
    }
}

File: 프로그래머스/3/43162. 네트워크/네트워크.cpp
#include <string>
#include <vector>

using namespace std;

// 컴퓨터 A와 컴퓨터 B 직접적 연결
// 컴퓨터 B와 컴퓨터 C 직접적 연결
// A-C도 정보교환 가능.
// A-B-C 같은 네트워크상에 있음.

// 네트워크 갯수는?
// computers[i][j] = 1이면 i,j연결. 

// dfs 구현. 

vector<int>graph[201];
bool visited[201] = {false};
int cnt = 0;

void dfs(int node){
    visited[node] = true;
        for (int g : graph[node]){
        if (!visited[g]) dfs(g);
    }
}

int solution(int n, vector<vector<int>> computers) {
    int answer = 0;
    
    for (int i=0; i<computers.size(); i++){
        for (int j=0; j<computers[i].size(); j++){
            if (computers[i][j] == 1 && i!=j){
                graph[i].push_back(j);
                graph[j].push_back(i);
            }
        }
    }
    
    for (int i=0; i<n; i++){
        if (!visited[i]) {
            dfs(i);
            cnt ++;
        }
    }
    
    return cnt;
}